WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController+Extensions.swift

```diff
@@ -241,6 +246,128 @@ extension NotificationsViewController: WPTableViewHandlerDelegate
 
 
 
+// MARK - Filter Helpers
+//
+extension NotificationsViewController
+{
+    func showFiltersSegmentedControlIfApplicable() {
+        guard tableHeaderView.alpha == WPAlphaZero && shouldDisplayFilters == true else {
+            return
+        }
+
+        UIView.animateWithDuration(WPAnimationDurationDefault) {
+            self.tableHeaderView.alpha = WPAlphaFull
+        }
+    }
+
+    func hideFiltersSegmentedControlIfApplicable() {
+        if tableHeaderView.alpha == WPAlphaFull && shouldDisplayFilters == false {
+            tableHeaderView.alpha = WPAlphaZero
+        }
+    }
+
+    var shouldDisplayFilters: Bool {
+        // Note:
+        // Filters should only be hidden whenever there are no Notifications in the bucket (contrary to the FRC's
+        // results, which are filtered by the active predicate!).
+        //
+        let bucket = simperium.bucketForName(entityName())
+        return bucket.numObjects() > 0
+    }
+}
+
+
+
+// MARK: - NoResults Helpers
+//
+extension NotificationsViewController
+{
+    func showNoResultsViewIfNeeded() {
+        // Remove + Show Filters, if needed
+        guard shouldDisplayNoResultsView == true else {
+            noResultsView.removeFromSuperview()
+            showFiltersSegmentedControlIfApplicable()
+            return
+        }
+
+        // Attach the view
+        if noResultsView.superview == nil {
+            tableView.addSubviewWithFadeAnimation(noResultsView)
+        }
+
+        // Refresh its properties: The user may have signed into WordPress.com
+        noResultsView.titleText     = noResultsTitleText
+        noResultsView.titleText     = noResultsTitleText
+        noResultsView.messageText   = noResultsMessageText
+        noResultsView.accessoryView = noResultsAccessoryView
+        noResultsView.buttonTitle   = noResultsButtonText
+
+        // Hide the filter header if we're showing the Jetpack prompt
+        hideFiltersSegmentedControlIfApplicable()
+    }
+
+    var noResultsTitleText: String {
+        guard shouldDisplayJetpackMessage == false else {
+            return NSLocalizedString("Connect to Jetpack", comment: "Notifications title displayed when a self-hosted user is not connected to Jetpack")
+        }
+
+        let messageMap: [Filter: String] = [
```

Comment:
Thanks James!! but i'll be iterating a bit over the `Filter` enum. I'm thinking about encapsulating the `noResults` string, and perhaps the NSPredicate format as well, in there. Since it's already tested, and i'll change this later, let's leave it this way for a bit.